### PR TITLE
fix unicode handling for swift and python 2 actions

### LINKED ISSUE #362: Swift action parsing doesn't tolerate special characters
For example, if I try to run an action with the following code:

``` swift
//  Copyright © 2016 Dirk Beanberger. All rights reserved.
func main(args: [String:Any]) -> [String:Any] {
  return ["message": "This should just work..."]
}
```

I always get an error of `the action did not produce a valid JSON response`

However, if I simply remove the `©` everything works as expected.

### PR DESCRIPTION
**Pull request by rabbah**
This replaces #2069 (was rebased on  pr #1903).
Fixes #362.
PG2/1327 🔵

### CODE REVIEW

File: tools/cli/go-whisk-cli/commands/action.go [hunk 1]
```diff
@@ -412,6 +412,11 @@ func getExec(artifact string, kind string, isDocker bool, mainEntry string) (*wh
         code, err = readFile(artifact)
         exec.Code = &code
 
+        if (ext == ".jar") {
```
Comment (rabbah) [inline tools/cli/go-whisk-cli/commands/action.go]: @dubeejw this fixes a bug in the CLI where --kind java doesn't encode the jar file.

File: core/actionProxy/actionproxy.py [hunk 2]
```diff
@@ -138,29 +141,35 @@ def error(msg):
...
-        if isinstance(o, bytes):
+        # version, so if bytes, decode to text. Note that in Python 2
+        # a string will match both types; so also skip decoding in that case
+        if isinstance(o, bytes) and not isinstance(o, str):
             o = o.decode('utf-8')
-        if isinstance(e, bytes):
+        if isinstance(e, bytes) and not isinstance(e, str):
             e = e.decode('utf-8')
 
         # get the last line of stdout, even if empty
```
Comment (rabbah) [inline core/actionProxy/actionproxy.py]: the change below is worth extra scrutiny.

**Comment by csantanapr** [comment]
@markusthoemmes could you help review this PR?

**Comment by rabbah** [comment]
@cbickel also a good reviewer for this if he has bandwidth.

**Comment by rabbah** [comment]
I suggest reviewing with https://github.com/openwhisk/openwhisk/pull/2070/files?w=1 to focus on actual changes which are small.

File: tools/travis/build.sh [hunk 3]
```diff
@@ -21,7 +21,7 @@ $ANSIBLE_CMD apigateway.yml
 
 cd $ROOTDIR
 
-./gradlew distDocker -x :core:swift3Action:distDocker -x :core:pythonAction:distDocker -PdockerImagePrefix=testing
```
Comment (csantanapr) [inline tools/travis/build.sh]: Why you remove this?
This excludes is to make the Travis build end below the 50 minutes max.
This related to the gradle task testLean, where swift3 and python are excluded 
https://github.com/openwhisk/openwhisk/blob/master/tests/build.gradle#L18

**Comment by csantanapr** [comment]
@rabbah 
>@cbickel also a good reviewer for this if he has bandwidth.

I added @cbickel to reviewers list

File: tools/travis/build.sh (hunk 3 above)
Comment (rabbah) [inline tools/travis/build.sh]: I looked at the build times and building the containers did not add enough time to get close to the limit and it was in the noise relative to other builds. We will now run sniff tests against all the runtimes in Travis which is better than contributors not knowing they've broken something systemic for the previously ignored runtimes.

We can further cut the swift build time by putting up a base image in dockerhub. We should we do as part of your work on all images published to docker hub.

Comment (csantanapr) [inline tools/travis/build.sh]: @rabbah if you are going to build the images, then it make sense to enable the tests in the build.gradle task testLean used in travis

The current tasks exclude swift and python tests.
```
task testLean(type: Test) {
    exclude '**/*Swift*'
    exclude '**/*Python*'
    exclude '**/*ThrottleTests*'
    exclude '**/MaxActionDurationTests*'
}
```

Comment (rabbah) [inline tools/travis/build.sh]: No change is needed here. The tests I enabled run as part of lean now.

Comment (csantanapr) [inline tools/travis/build.sh]: I don't understand what you mean, but OK
 you didn't modified gradle.build and travis still calls `./gradlew :tests:testLean`

Comment (rabbah) [inline tools/travis/build.sh]: Inspect this line: https://github.com/openwhisk/openwhisk/pull/2070/files#diff-31a3b6664356184ce7bd3d6d5d6ef4d0R39

This test suite is already running (it's included in `testLean` already).

File: tests/src/test/scala/system/basic/WskUnicodeTests.scala [hunk 4]
```diff
@@ -36,16 +36,19 @@ class WskUnicodeTests
...
+                    val name = s"unicodeGalore.${k.replace(":", "")}"
 
                     assetHelper.withCleaner(wsk.action, name) {
                         (action, _) =>
-                            action.create(name, Some(TestUtils.getTestActionFilename(file)), main = if (k == "java") Some("Unicode") else None)
+                            action.create(
+                                name,
+                                Some(TestUtils.getTestActionFilename(file)),
+                                main = if (k == "java") Some("Unicode") else None,
+                                kind = Some(k))
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskUnicodeTests.scala]: @dubeejw here's the test that surfaced the bug mentioned below.

**Comment by rabbah** [comment]
thanks @cbickel for having a look.

File: tools/cli/go-whisk-cli/commands/action.go (hunk 1 above)
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: The CLI change LGTM.

Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: I think we might be able to Base64 encode here for .jar and .zip files. And remove the encoding for .zip files that is at the end of the function.

```
    if ext == “.jar” || ext == “.zip”  {
        code = base64.StdEncoding.EncodeToString([]byte(code))
        exec.Code = &code
    }

```

Comment (rabbah) [inline tools/cli/go-whisk-cli/commands/action.go]: good call. https://github.com/openwhisk/openwhisk/pull/2070/commits/0e0cc206a499719794d7f0e6024a1b19a6bdf4c9

File: tools/cli/go-whisk-cli/commands/action.go [hunk 5]
```diff
@@ -410,11 +410,18 @@ func getExec(artifact string, kind string, isDocker bool, mainEntry string) (*wh
...
-        if (ext == ".jar") {
-            exec.Jar = base64.StdEncoding.EncodeToString([]byte(code))
-            exec.Code = nil
+        if ext == ".zip" || ext == ".jar" {
+            // Base64 encode the file
+            code = base64.StdEncoding.EncodeToString([]byte(code))
+            if (ext == ".zip") {
+                exec.Code = &code
+            } else {
+                exec.Code = nil
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: Is the assignment of nil necessary here?

Comment (rabbah) [inline tools/cli/go-whisk-cli/commands/action.go]: https://github.com/openwhisk/openwhisk/pull/2070/files#diff-41c5935cb68718d73fa6ef611c927cc4L439

Comment (rabbah) [inline tools/cli/go-whisk-cli/commands/action.go]: It seems now that since exec.Code is not already assigned, we can remove this.

Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: The way to tell would be to create a action with a .jar file and make sure the `code` field is not present in the package sent from the CLI.

Comment (rabbah) [inline tools/cli/go-whisk-cli/commands/action.go]: I removed `Jar` entirely.

**Comment by rabbah** [comment]
@bjustin-ibm this is now rebased to master and fully squashed.

**Comment by bjustin-ibm** [comment]
PG2 1376 🔵 (for good measure)

**Pull request merged**